File: src/poll_agents/repository/supabase.py
"""Supabase repository implementation."""

from datetime import datetime
from typing import Optional

from supabase import create_client, Client

from ..models import QuestionSet, AgentResponse
from .base import QuestionSetRepository, ResponseRepository


class SupabaseClientManager:
    """Manages Supabase client lifecycle."""

    def __init__(self, url: str, key: str):
        self.url = url
        self.key = key
        self._client: Optional[Client] = None

    def get_client(self) -> Client:
        """Get or create Supabase client."""
        if self._client is None:
            self._client = create_client(self.url, self.key)
        return self._client


class SupabaseQuestionSetRepository(QuestionSetRepository):
    """Supabase-backed question set repository."""

    def __init__(self, client_manager: SupabaseClientManager):
        self.client_manager = client_manager

    def _to_model(self, data: dict) -> QuestionSet:
        """Convert Supabase row to QuestionSet model."""
        created_at = data["created_at"]
        if isinstance(created_at, str):
            created_at = datetime.fromisoformat(created_at.replace("Z", "+00:00"))
        return QuestionSet(
            id=data["id"],
            name=data["name"],
            q1=data["q1"],
            q2=data["q2"],
            q3=data["q3"],
            created_at=created_at,
            active=data.get("active", True),
        )

    async def get_active(self) -> Optional[QuestionSet]:
        """Get the most recently created active question set."""
        client = self.client_manager.get_client()
        response = (
            client.table("question_sets")
            .select("*")
            .eq("active", True)
            .order("created_at", desc=True)
            .limit(1)
            .execute()
        )
        if response.data:
            return self._to_model(response.data[0])
        return None

    async def get_by_id(self, id: str) -> Optional[QuestionSet]:
        """Get a question set by ID."""
        client = self.client_manager.get_client()
        response = client.table("question_sets").select("*").eq("id", id).limit(1).execute()
        if response.data:
            return self._to_model(response.data[0])
        return None

    async def create(self, question_set: QuestionSet) -> None:
        """Create a new question set."""
        client = self.client_manager.get_client()
        client.table("question_sets").insert({
            "id": question_set.id,
            "name": question_set.name,
            "q1": question_set.q1,
            "q2": question_set.q2,
            "q3": question_set.q3,
            "created_at": question_set.created_at.isoformat(),
            "active": question_set.active,
        }).execute()

    async def get_all(self) -> list[QuestionSet]:
        """Get all question sets."""
        client = self.client_manager.get_client()
        response = (
            client.table("question_sets")
            .select("*")
            .order("created_at", desc=True)
            .execute()
        )
        return [self._to_model(item) for item in response.data]


class SupabaseResponseRepository(ResponseRepository):
    """Supabase-backed response repository."""

    def __init__(self, client_manager: SupabaseClientManager):
        self.client_manager = client_manager

    def _to_model(self, data: dict) -> AgentResponse:
        """Convert Supabase row to AgentResponse model."""
        completed_at = data["completed_at"]
        if isinstance(completed_at, str):
            completed_at = datetime.fromisoformat(completed_at.replace("Z", "+00:00"))
        return AgentResponse(
            id=data["id"],
            question_set_id=data["question_set_id"],
            agent_email=data["agent_email"],
            a1=data["a1"],
            a2=data["a2"],
            a3=data["a3"],
            completed_at=completed_at,
        )

    async def save(self, response: AgentResponse) -> None:
        """Save an agent response."""
        client = self.client_manager.get_client()
        client.table("agent_responses").insert({
            "id": response.id,
            "question_set_id": response.question_set_id,
            "agent_email": response.agent_email,
            "a1": response.a1,
            "a2": response.a2,
            "a3": response.a3,
            "completed_at": response.completed_at.isoformat(),
        }).execute()

        # Print to console
        print("\n" + "=" * 50)
        print("NEW AGENT RESPONSE RECORDED (Supabase)")
        print("=" * 50)
        print(f"Email: {response.agent_email}")
        print(f"Question Set: {response.question_set_id}")
        print(f"Answers: {['Yes' if a else 'No' for a in response.answers]}")
        print(f"Completed: {response.completed_at}")
        print("=" * 50 + "\n")

    async def get_by_email(self, email: str) -> list[AgentResponse]:
        """Get all responses from an agent by email."""
        client = self.client_manager.get_client()
        response = client.table("agent_responses").select("*").eq("agent_email", email).execute()
        return [self._to_model(item) for item in response.data]

File: src/repository/supabase.ts
/** Supabase repository implementation. */

import { createClient, type SupabaseClient } from "@supabase/supabase-js";
import type { QuestionSet, AgentResponse } from "../models";
import { getAnswers } from "../models";
import type { QuestionSetRepository, ResponseRepository } from "./base";

export class SupabaseClientManager {
  private client: SupabaseClient | null = null;

  constructor(
    private url: string,
    private key: string,
  ) {}

  getClient(): SupabaseClient {
    if (!this.client) {
      this.client = createClient(this.url, this.key);
    }
    return this.client;
  }
}

function toQuestionSet(data: Record<string, unknown>): QuestionSet {
  const createdAt =
    typeof data.created_at === "string"
      ? new Date(data.created_at)
      : (data.created_at as Date);
  return {
    id: data.id as string,
    name: data.name as string,
    q1: data.q1 as string,
    q2: data.q2 as string,
    q3: data.q3 as string,
    created_at: createdAt,
    active: (data.active as boolean) ?? true,
  };
}

function toAgentResponse(data: Record<string, unknown>): AgentResponse {
  const completedAt =
    typeof data.completed_at === "string"
      ? new Date(data.completed_at)
      : (data.completed_at as Date);
  return {
    id: data.id as string,
    question_set_id: data.question_set_id as string,
    agent_email: data.agent_email as string,
    a1: data.a1 as boolean,
    a2: data.a2 as boolean,
    a3: data.a3 as boolean,
    completed_at: completedAt,
  };
}

export class SupabaseQuestionSetRepository implements QuestionSetRepository {
  constructor(private clientManager: SupabaseClientManager) {}

  async getActive(): Promise<QuestionSet | null> {
    const client = this.clientManager.getClient();
    const { data, error } = await client
      .from("question_sets")
      .select("*")
      .eq("active", true)
      .order("created_at", { ascending: false })
      .limit(1);
    if (error) throw error;
    if (data && data.length > 0) return toQuestionSet(data[0]);
    return null;
  }

  async getById(id: string): Promise<QuestionSet | null> {
    const client = this.clientManager.getClient();
    const { data, error } = await client
      .from("question_sets")
      .select("*")
      .eq("id", id)
      .limit(1);
    if (error) throw error;
    if (data && data.length > 0) return toQuestionSet(data[0]);
    return null;
  }

  async create(questionSet: QuestionSet): Promise<void> {
    const client = this.clientManager.getClient();

    // If new set is active, deactivate all others first
    if (questionSet.active) {
      const { error: deactivateError } = await client
        .from("question_sets")
        .update({ active: false })
        .eq("active", true);
      if (deactivateError) throw deactivateError;
    }

    const { error } = await client.from("question_sets").insert({
      id: questionSet.id,
      name: questionSet.name,
      q1: questionSet.q1,
      q2: questionSet.q2,
      q3: questionSet.q3,
      created_at: questionSet.created_at.toISOString(),
      active: questionSet.active,
    });
    if (error) throw error;
  }

  async setActive(id: string): Promise<void> {
    const client = this.clientManager.getClient();

    // Deactivate all question sets
    const { error: deactivateError } = await client
      .from("question_sets")
      .update({ active: false })
      .eq("active", true);
    if (deactivateError) throw deactivateError;

    // Activate the specified one
    const { error: activateError } = await client
      .from("question_sets")
      .update({ active: true })
      .eq("id", id);
    if (activateError) throw activateError;
  }

  async getAll(): Promise<QuestionSet[]> {
    const client = this.clientManager.getClient();
    const { data, error } = await client
      .from("question_sets")
      .select("*")
      .order("created_at", { ascending: false });
    if (error) throw error;
    return (data ?? []).map(toQuestionSet);
  }
}

export class SupabaseResponseRepository implements ResponseRepository {
  constructor(private clientManager: SupabaseClientManager) {}

  async save(response: AgentResponse): Promise<void> {
    const client = this.clientManager.getClient();
    const { error } = await client.from("agent_responses").insert({
      id: response.id,
      question_set_id: response.question_set_id,
      agent_email: response.agent_email,
      a1: response.a1,
      a2: response.a2,
      a3: response.a3,
      completed_at: response.completed_at.toISOString(),
    });
    if (error) throw error;

    const answers = getAnswers(response);
    console.log("");
    console.log("=".repeat(50));
    console.log("NEW AGENT RESPONSE RECORDED (Supabase)");
    console.log("=".repeat(50));
    console.log(`Email: ${response.agent_email}`);
    console.log(`Question Set: ${response.question_set_id}`);
    console.log(
      `Answers: [${answers.map((a) => (a ? "y" : "n")).join(", ")}]`,
    );
    console.log(`Completed: ${response.completed_at.toISOString()}`);
    console.log("=".repeat(50));
    console.log("");
  }

  async getByEmail(email: string): Promise<AgentResponse[]> {
    const client = this.clientManager.getClient();
    const { data, error } = await client
      .from("agent_responses")
      .select("*")
      .eq("agent_email", email);
    if (error) throw error;
    return (data ?? []).map(toAgentResponse);
  }

  async hasResponded(email: string, questionSetId: string): Promise<boolean> {
    const client = this.clientManager.getClient();
    const { data, error } = await client
      .from("agent_responses")
      .select("id")
      .eq("agent_email", email)
      .eq("question_set_id", questionSetId)
      .limit(1);
    if (error) throw error;
    return (data?.length ?? 0) > 0;
  }
}
